1.MLP_Binary_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,627 @@
+        print(f"Validation Loss: {np.mean(valid_loss)}")
+        print(f"Training Accuracy: {np.mean(train_acc)}")
+        print(f"Training Loss: {np.mean(train_loss)}")
+    else:
+        return np.mean(valid_loss)
+
+
+def extract_best_trial_params(best_params):
+    """
+    This function extractions the best parameters from the best trial.
+
+    These extracted parameters will be used to create a new model.
+
+    Parameters:
+        best_params : obj
+            study.best_params
+            the best_params function of the study object
+    Return:
+        param_dict : dict
+            dictionary of all of the params for the best trial model
+    """
+
+    params = best_params
+    units = []
+    dropout = []
+    n_layers = params["n_layers"]
+    optimizer = params["optimizer"]
+    lr = params["learning_rate"]
+    for i in range(params["n_layers"]):
+        units.append(params[f"n_units_l{i}"])
+        dropout.append(params[f"dropout_{i}"])
+    param_dict = {
+        "units": units,
+        "dropout": dropout,
+        "n_layers": n_layers,
+        "optimizer": optimizer,
+        "learning_rate": lr,
+    }
+    return param_dict
+
+
+# function for new optimized model
+def optimized_model(in_features, paramater_dict):
+    """
+    This function uses the extracted optimized functions to create a new model
+
+    Parameters:
+        in_features : int
+            this is the number of in features to used for the model
+        parameter_dict : dict
+            this is a dictionary returned fropm the extract_best_trial_params function
+
+    Return:
+        nn.Sequential(*layers) : dict
+            this returns in a dict the architecture of the model with optimized parameters
+    """
+    n_layers = paramater_dict["n_layers"]
+
+    layers = []
+
+    in_features
+
+    for i in range(n_layers):
+
+        out_features = paramater_dict["units"][i]
+
+        layers.append(nn.Linear(in_features, out_features))
+        layers.append(nn.ReLU())
+        p = paramater_dict["dropout"][i]
+        layers.append(nn.Dropout(p))
+        in_features = out_features
+    layers.append(nn.Linear(out_features, 1))
+
+    return nn.Sequential(*layers)
+
+
+# Model Training
+def train_optimized_model(
+    EPOCHS,
+    train_loader,
+    valid_loader,
+    in_features,
+    parameter_dict,
+    params,
+    return_info=False,
+):
+    """
+    This function trains the optimized model on the training dataset
+
+    Parameters:
+        EPOCHS : int
+            the number of epochs to train the model for
+    Return:
+        training metrics : str
+
+    """
+
+    model = optimized_model(in_features, parameter_dict)
+    model = model.to(params.DEVICE)
+    criterion = nn.BCEWithLogitsLoss()
+    optim_method = parameter_dict["optimizer"].strip("'")
+    optimizer = (
+        f'optim.{optim_method}(model.parameters(), lr=parameter_dict["learning_rate"])'
+    )
+    optimizer = eval(optimizer)
```

Comment:
(Reposting from other conversation, again apologies for the confusion on this one). In some circumstances, eval is considered an insecure implementation. While I don't have concerns about security here, it's good practice to avoid using it if possible. Consider using [ast.literal_eval](https://docs.python.org/3/library/ast.html#ast.literal_eval) instead.

Reply:
I am having issues getting `ast.literal_eval` to work we might need to sidebar at some point to talk through this.